File: Sudoku Solver/sudoku.py
import pandas as pd 
import numpy as np 

UNASSIGNED = 0
rank = 3
grid = [[5, 3, 0, 0, 7, 0, 0, 0, 0],
        [6, 0, 0, 1, 9, 5, 0, 0, 0],
        [0, 9, 8, 0, 0, 0, 0, 6, 0],
        [8, 0, 0, 0, 6, 0, 0, 0, 3],
        [4, 0, 0, 8, 0, 3, 0, 0, 1],
        [7, 0, 0, 0, 2, 0, 0, 0, 6],
        [0, 6, 0, 0, 0, 0, 2, 8, 0],
        [0, 0, 0, 4, 1, 9, 0, 0, 5],
        [0, 0, 0, 0, 8, 0, 0, 7, 9]]
#sudoku = np.matrix(grid)

def isValid(self,row, col, number):
    global grid 
    global rank
    print("Valid?")

    if(number != UNASSIGNED):
        for i in range(0, rank * rank, 1):
            if(grid[row][i] == number and i != col):
                return False

        for j in range(0, rank * rank, 1):
            if(grid[i][col] == number and i != row):
                return False

    #Für die Kästchen
    r = (row  // rank) * rank #Für die Zeile
    c = (col // rank) * rank #Für die Spalte

    for i in range(0,rank):
        for j in range(0,rank):
            if(grid[i+r][j+c] == number and row != i + r and col != j + c):
                return False

    return True

def solveSudoku():
    global grid
    global rank

    for row in range(rank * rank):
        for col in range(rank * rank):
            if(grid[row][col] == UNASSIGNED):
                for number in range(1, (rank * rank) + 1):
                    if isValid(row, col, number):
                        grid[row][col] = number
                        solveSudoku()
                        grid[row][col] = UNASSIGNED
                return
    print(np.matrix(grid))
    input("More?")




def solve(self,rank):
    global grid
    self.rank = rank
    
    for i in range(rank * rank):
        for j in range(rank * rank):
            field = grid[i][j]

            if(not(isValid(i,j,field))):
                print("Sudoku isnt valid!")
                exit()
    if(rank < 1):
        print("The Rank has to be min. 2")
        exit()

    solveSudoku()

solveSudoku()
# def __main__():
#     solveSudoku()
#    # solve(3)

File: Sudoku Solver/sudoku.java
public class SeriousSudokuSolver {
	private static int[][] mySudoku;
	private static final int UNASSIGNED = -1;
	private static int n;

	// Guckt ob ein Eintrag valid wäre
	private static boolean isValid(int row, int col, int number) {
		if (number != UNASSIGNED) {

			// Für die Zeile
			for (int i = 0; i < n * n; i++) {
				if (mySudoku[row][i] == number && i != col) {
					return false;
				}
			}

			// Für die Spalte
			for (int i = 0; i < n * n; i++) {
				if (mySudoku[i][col] == number && i != row) {
					return false;
				}
			}

			// Fürs Kästchen
			int r = row - row % n; // Für die Zeile
			int c = col - col % n; // Für die Spalte

			for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					if (mySudoku[i + r][j + c] == number && row != i + r && col != j + c) {
						return false;
					}
				}
			}

			return true;
		}

		return true;
	}

	public static boolean solveSudoku() {

		for (int row = 0; row < n * n; row++) {
			for (int col = 0; col < n * n; col++) {

				if (mySudoku[row][col] == UNASSIGNED) {

					for (int number = 1; number <= n * n; number++) {
						mySudoku[row][col] = number;
						if (!isValid(row, col, number)) {
							mySudoku[row][col] = UNASSIGNED;
						} else {
							if (solveSudoku()) {
								return true;
							} else {
								mySudoku[row][col] = UNASSIGNED;
							}
						}
						// if (isValid(row, col, number)) {
						// mySudoku[row][col] = number;
						//
						// if (solveSudoku()) {
						// return true;
						// } else {
						// mySudoku[row][col] = UNASSIGNED;
						// }
						// }
					}
					return false;
				}
			}
		}
		return true;
	}

	/**
	 * Solves the given Sudoku through backtracking.
	 * 
	 * @param rank
	 *            Determines the size of the Sudoku. A Field of rank n will have n^2
	 *            columns and rows.
	 * @param sudoku
	 *            The start configuration of the Sudoku. Empty cells are initialized
	 *            as -1.
	 * @return The solved Sudoku.
	 * @throws IllegalArgumentException
	 *             if the given Sudoku is impossible to solve.
	 */
	public static int[][] solve(int rank, int[][] sudoku) throws IllegalArgumentException {
		n = rank;
		mySudoku = sudoku;

		// Überprüfun ob gegebenes Sudoku überhaupt valide ist
		for (int i = 0; i < n * n; i++) {
			for (int j = 0; j < n * n; j++) {
				int field = sudoku[i][j];

				if (!isValid(i, j, field)) {
					throw new IllegalArgumentException("Sudoku ist nicht valide.");
				}

			}
		}

		if (rank < 1) {
			throw new IllegalArgumentException("Das Sudoku muss mindestens 1 groß sein");
		} else {
			if (!solveSudoku()) {
				throw new IllegalArgumentException("Sudoku ist nicht lösbar.");
			} else {
				return mySudoku;
			}
		}

	}
}